scripts/macos/RenameMacUserNameAndHomeDirectory.sh

```diff
@@ -1,61 +1,137 @@
 fi
 
-dscl . -change /Users/${oldUser} NFSHomeDirectory /Users/${oldUser} /Users/${newUser}
+#Checks if user is logged in
+killProcess=$(ps -Ajc | grep $oldUser | grep loginwindow | awk '{print $2}')
+
+#Logs out user if they are logged in by killing their loginwindow process
+if [[ "${killProcess}" ]]; then
+	sudo kill -9 $killProcess
+	echo "$oldUser account was logged in. User logged off to complete username update."
+fi
+
+#Captures current "RealName" this is the displayName
+fullRealName=$(dscl . -read /Users/${oldUser} RealName)
+
+#Formats "RealName"
+origRealName=$(echo $fullRealName | cut -d' ' -f2-)
+
+#Updates "RealName" to new username (Yes JCAgent will overwrite this after user/system association)
+sudo dscl . -change "/Users/${oldUser}" RealName "${origRealName}" "${newUser}"
+
+err=$?
+if [[ ${err} -ne 0 ]]; then
+	echo "Could not rename the user's RealName in dscl. - err=${err}"
+	exit 1
+fi
+
+#Captures current NFS home directory
+origHomeDir=$(dscl . -read "/Users/${oldUser}" NFSHomeDirectory | awk '{print $2}' -)
+
+if [[ -z "${origHomeDir}" ]]; then
+	echo "Cannot obtain the original home directory name, is the oldUserName correct?"
+	exit 1
+fi
+
+#Updates NFS home directory
+sudo dscl . -change "/Users/${oldUser}" NFSHomeDirectory "${origHomeDir}" "/Users/${newUser}"
+
 err=$?
-if [ ${err} -ne 0 ]
-then
-	abort "Could not rename the user's home directory pointer, aborting further changes! - err=${err}"
+if [ ${err} -ne 0 ]; then
+	echo "Could not rename the user's home directory pointer, aborting further changes! - err=${err}"
+	exit 1
 fi
 
-mv /Users/${oldUser} /Users/${newUser}
+#Updates name of home directory to new username
+mv "${origHomeDir}" "/Users/${newUser}"
+
 err=$?
-if [[ ${err} -ne 0 ]]
-then
-	abort "Could not rename the user's home directory in /Users - the user may not be able to login unless you correct dscl to point back to /Users/${oldUser}"
+if [[ ${err} -ne 0 ]]; then
+	echo "Could not rename the user's home directory in /Users - the user may not be able to login unless you correct dscl to point back to /Users/${oldUser}"
+	exit 1
 fi
 
-dscl . -change /Users/${oldUser} RecordName ${oldUser} ${newUser}
+#Actual username change
+sudo dscl . -change "/Users/${oldUser}" RecordName "${oldUser}" "${newUser}"
+
 err=$?
-if [[ ${err} -ne 0 ]]
-then
-	abort "Could not rename the user's RecordName in dscl - the user should still be able to login, but with user name ${oldUser}, however, their home directory will be pointed to /Users/${newUser}"
+if [[ ${err} -ne 0 ]]; then
+	echo "Could not rename the user's RecordName in dscl - the user should still be able to login, but with user name ${oldUser}, however, their home directory will be pointed to /Users/${newUser}"
+	exit 1
 fi
 
+#Links old home directory to new. Fixes dock breaking issue
+ln -s "/Users/${newUser}" "/Users/${oldUser}"
+
+#Restarts the Mac menu bar to show the updated username in the fast user switching menu
+killall -KILL SystemUIServer
```

Comment:
I cringe whenever I see a `-KILL` | `-9` signal being used, unless as a last resort. Probably because I've seen many times where that has left to unclean processes. Is there a more _polite_ way to restart the menu bar?

Reply:
@dbyington 
I updated the brackets and found a nicer way to log off the user not using -kill 9... I'm searching for a way to restart the menu bar "nicer" Do you think `killall SystemUIServer` is better than `killall -KILL SystemUIServer`? I've run both commands locally and don't notice a difference.

Reply:
Since it appears Apple doesn't provide a way to osascript a message to the `SystemUIServer` to reload/restart I'd probably go with a `SIGHUP`. The `HUP` is the typical signal used to tell something to reload or restart. It appears to work nicely with `SystemUIServer`.